airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduper.java

```diff
@@ -0,0 +1,137 @@
+ * exception if that method was not called.
+ */
+public class DefaultTyperDeduper<DialectTableDefinition> implements TyperDeduper {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(TyperDeduper.class);
+
+  private static final String NO_SUFFIX = "";
+  private static final String TMP_OVERWRITE_TABLE_SUFFIX = "_airbyte_tmp";
+
+  private final SqlGenerator<DialectTableDefinition> sqlGenerator;
+  private final DestinationHandler<DialectTableDefinition> destinationHandler;
+  private final ParsedCatalog parsedCatalog;
+  private Map<StreamId, String> overwriteStreamsWithTmpTable;
+
+  public DefaultTyperDeduper(SqlGenerator<DialectTableDefinition> sqlGenerator,
+                             DestinationHandler<DialectTableDefinition> destinationHandler,
+                             ParsedCatalog parsedCatalog) {
+    this.sqlGenerator = sqlGenerator;
+    this.destinationHandler = destinationHandler;
+    this.parsedCatalog = parsedCatalog;
+  }
+
+  /**
+   * Create the tables that T+D will write to during the sync. In OVERWRITE mode, these might not be
+   * the true final tables. Specifically, other than an initial sync (i.e. table does not exist, or is
+   * empty) we write to a temporary final table, and swap it into the true final table at the end of
+   * the sync. This is to prevent user downtime during a sync.
+   */
+  public void prepareFinalTables() throws Exception {
+    if (overwriteStreamsWithTmpTable != null) {
+      throw new IllegalStateException("Tables were already prepared.");
+    }
+    overwriteStreamsWithTmpTable = new HashMap<>();
+
+    // For each stream, make sure that its corresponding final table exists.
+    // Also, for OVERWRITE streams, decide if we're writing directly to the final table, or into an
+    // _airbyte_tmp table.
+    for (StreamConfig stream : parsedCatalog.streams()) {
+      final Optional<DialectTableDefinition> existingTable = destinationHandler.findExistingTable(stream.id());
+      if (existingTable.isEmpty()) {
+        // The table doesn't exist. Create it.
+        destinationHandler.execute(sqlGenerator.createTable(stream, NO_SUFFIX));
+        if (stream.destinationSyncMode() == DestinationSyncMode.OVERWRITE) {
+          // We're creating this table for the first time. Write directly into it.
+          overwriteStreamsWithTmpTable.put(stream.id(), NO_SUFFIX);
+        }
+      } else {
+        // The table exists. Make sure it has the right schema.
+        if (!sqlGenerator.existingSchemaMatchesStreamConfig(stream, existingTable.get())) {
+          LOGGER.info("Existing schema for stream {} is different from expected schema. Executing soft reset.", stream.id().finalTableId(""));
+          for (String sql : sqlGenerator.softReset(stream)) {
+            destinationHandler.execute(sql);
+          }
+        } else {
+          LOGGER.info("Existing schema for stream {} matches expected schema, no alterations needed", stream.id().finalTableId(""));
+        }
```

Comment:
it wasn't... but I think this logic and the previous logic are both wrong 🤦 can you doublecheck me here?

joe's logic:
1. if we're non-overwrite, then we should check the existing table schema + do a soft reset
2. if we're overwrite, then do nothing

Which I think is wrong. If we're in overwrite mode but the existing table is empty, then we'll T+D directly into it - which will fail because the existing table has the wrong schema.

my logic:
1. Always check the existing table schema + do a soft reset

Which is technically correct, but wasteful - in overwrite mode when we're writing to a tmp table, we would unnecessarily rebuild the true final table, even though we're already planning to replace it at the end of the sync.

new logic that I just wrote:
1. If we're non-overwrite, then we should check the existing table schema + do a soft reset
2. If we're overwrite and writing to the existing table directly, then we should check schema + soft reset
3. If we're overwrite and writing to a tmp table, then we don't need to check the schema

Which I think solves both those problems, without introducing new problems?

Reply:
yes I think this makes sense 👍